Chunked w-gridder with dask-ms and ducc

landmanbester opened the pull request:
I am trying to build a chunked up interface for the wgridder using dask-ms. I am mainly thinking about the functionality I need for pfb-clean but let's discuss here @sjperkins @o-smirnov @mreineck.  

My thinking is to traverse the chunks in serial and leave all the parallelisation to the ducc.wgridder (derived from the nifty_gridder). This should be better for memory consumption and performance but let's see (interleaved io could be very cool here indeed). Anyway, I seem to have a working gridder (the degridder is tomorrow's task). I really expected this to be harder as there is a certain amount of nested parallelism involved. However, to my surprise, setting (so that the chunks are traversed in serial)

dask.config.set(pool=ThreadPool(1))

and passing in nthreads > 1 parameter to the wgridder, magically spins up all threads to 100% with all of them looking beautifully green. The dirty image looks like what I expect but I am yet to test that it agrees with the non-dask version. I would love some feedback on this. Please check out the docstrings in this class

https://github.com/landmanbester/codex-africanus/blob/f973bb149b03cf5652f6a6cb838d9327b6d2b4c9/africanus/gridding/nifty/dask.py#L94

where I have tried to describe the intended behaviour.

sjperkins commented:
@landmanbester I'll wait for you to implement the degridding functionality before taking a closer look.

Regarding the nested parallelism, I can see this is going to be a topic that needs be addressed in a paradigm where we're running code using OpenMP/TBB parallelism inside dask threads.

Playing the devil's advocate: too many dask + OpenMP threads lead to [oversubscription](https://docs.microsoft.com/en-us/visualstudio/profiling/common-patterns-for-poorly-behaved-multithreaded-applications?view=vs-2019), while if we only use a single dask thread, then what's the point of using dask?

The answer may lie in the middle. In past conversations with @JSKenyon, he mentioned a hybrid strategy of some dask threads and some OpenMP/TBB threads working well. Perhaps the same might apply to the NIFTY gridder? Some support for this kind of parallelism was implemented for the current NIFTY dask wrappers (below the new code you've added) where the user can ask for N gridding streams.

This is not quite `ThreadPool(N)`: it creates the graph in such a way that there are N serial gridding chains rather than gridding each visibility chunk separately and summing them together in a [Parallel Reduction](https://en.wikipedia.org/wiki/Reduce_%28parallel_pattern%29)

If [Task Annotations](https://github.com/dask/dask/pull/6217)  makes it into dask, we'll have finer control of the scheduling of tasks. For example: creating a dedicated gridding thread or dedicated solver thread per dask worker on which only these operations may run.

mreineck commented:
It most likely makes no difference, but I'd still like to point this out: the threads used within DUCC are plain POSIX threads, not OpenMP or TBB. Just so you are not surprised in case you are analysing the code with OpenMP-specific tools and don't find any parallelism ;-)

sjperkins commented:
> It most likely makes no difference, but I'd still like to point this out: the threads used within DUCC are plain POSIX threads, not OpenMP or TBB. Just so you are not surprised in case you are analysing the code with OpenMP-specific tools and don't find any parallelism ;-)

Ah I see you've implemented your own ThreadPools and Schedulers in ducc0.

mreineck commented:
I was using OpenMP originally and was quite happy with it. However, when `pypocketfft` was integrated into `scipy`, we needed to ditch OpenMP since it does not always interact well with Python's own multithreading (no idea which side is to blame here). Peter Bell, who has helped me a lot with `pypocketfft` provided a pthread-based implementation which worked quite well, but originally only had "static" scheduling.

Later on I saw Andre Offringa's approach to multi-threading in `wsclean`, and from those two starting points I managed to put something together which is almost as powerful as OpenMP, but relies only on standard C++ features. I'm now using it everywhere and so far didn't encounter any problems.

landmanbester commented:
Well, this is why I want to discuss the implementation here. Currently, we are mainly chunking up the data because it doesn't fit into memory. Doesn't a parallel reduction always come with a linear increase in memory? This is exactly what I want to avoid.

My reasons for using dask are twofold:
i) dask-ms is a very neat and useful interface to the measurement set that handles chunking in a way that makes the code far more transparent than it would be if there was clunky hand-written chunking logic involved. Truth be told I don't even know how to achieve the same thing using pyrap directly. 
ii) I really hope we are moving towards a distributed gridding implementation that completely avoids the disk (as would be required for resolve type algorithms) and I think dask distributed is the way to go.

Anyway, the wrapper around the degridder is doing something but it's not doing it correctly. I see threads spinning up and data does end up in the measurement set but it also grows by about a factor of 6 after running the predict. I suspect the logic here

https://github.com/landmanbester/codex-africanus/blob/b67a29229f1dd36bcf4a8f25cf8489ef8efabd7c/africanus/gridding/nifty/dask.py#L400

@sjperkins will appreciate it if you can have a look when you get a chance

sjperkins commented:
> Well, this is why I want to discuss the implementation here. Currently, we are mainly chunking up the data because it doesn't fit into memory. Doesn't a parallel reduction always come with a linear increase in memory? This is exactly what I want to avoid.

It seems to be linear in the number of threads, at maximum. In fact, if a parallel reduction was used to sum the dirties produced by each chunk, I'd expect to see a memory profile similar to that of shadems which also use a parallel reduction: https://github.com/holoviz/datashader/pull/926. In particular see this comment: https://github.com/ratt-ru/shadeMS/issues/34#issuecomment-637676504 and https://github.com/ratt-ru/shadeMS/issues/34#issuecomment-638785524. shadems plots are 1024\*\*2, whereas it sounds like MeerKAT-sized grids are 8192\*\*2. (but what about the internal w-planes ducc uses internally?)

The shadems images were part of 3D cube with an image for each category, so I'm not even sure the data sizes are that different from each other. In any case, the visibility data dominated in my admittedly pen and paper exercises.

I also want to very much emphasise that shadems and parallel gridding are doing exactly the same thing in terms of the distributed algorithm, except shadems is plotting vis data, while here it is gridded. Additionally, shadems must support more reduction cases because it supports plotting every axis against most other axes (Correct me here @o-smirnov).

> My reasons for using dask are twofold:
> i) dask-ms is a very neat and useful interface to the measurement set that handles chunking in a way that makes the code far more transparent than it would be if there was clunky hand-written chunking logic involved. Truth be told I don't even know how to achieve the same thing using pyrap directly.

Cool, the intention has been for dask-ms to be an easy interface for the algorithm writer. I think that dask Arrays are also easy for the algorithm writer to work with.

> ii) I really hope we are moving towards a distributed gridding implementation that completely avoids the disk (as would be required for resolve type algorithms) and I think dask distributed is the way to go.

I agree. This is currently doable with the distributed Client interface, but it requires the user to manually place tasks on workers, rather than depending on the distributed scheduler to "do the right thing". As discussed in https://github.com/dask/dask/pull/6217#issuecomment-631088645 I believe the the more general approach is to [annotate](https://github.com/dask/dask/pull/6217) individual dask tasks with information that scheduler plugins (written by us) use to decide where to place tasks.

> 
> Anyway, the wrapper around the degridder is doing something but it's not doing it correctly. I see threads spinning up and data does end up in the measurement set but it also grows by about a factor of 6 after running the predict. I suspect the logic here
> 
> https://github.com/landmanbester/codex-africanus/blob/b67a29229f1dd36bcf4a8f25cf8489ef8efabd7c/africanus/gridding/nifty/dask.py#L400
> 
> @sjperkins will appreciate it if you can have a look when you get a chance

Sure, will take a look.

landmanbester commented:
As far as I know, the gridder's memory requirements do not scale with the number of w-stacks (please correct me if I am wrong, @mreineck). @sjperkins, if you have suggestions on how to improve the wrapper I am all ears. My basic design principles are:

i) the disk in unavoidable when it comes the predict since the data don't fit into memory. But we can assume that the image does.
ii) I need an operator that produces a residual image and a chi-square value by touching the disk as little as possible. I don't need to save these intermediate visibilities to disk (in place subtraction, read but no write) and I need to apply this operator at least O(10) times.
iii) I also need a way to write MODEL_DATA and RESIDUAL_VIS to the MS but I only need to do that once so speed is less of an issue

mreineck commented:
> As far as I know, the gridder's memory requirements do not scale with the number of w-stacks (please correct me if I am wrong, @mreineck).

That's correct.

sjperkins commented:
@landmanbester  Just to recap our zoom conversation:

1. The paradigm we're working with is that dask Arrays are the interface between the data source/sink and the algorithm.
2. For this reason, we shouldn't explicitly include dask-ms logic within codex-africanus algorithms.
3. I'm happy to explore the idea of a wgridder object with dot and hdot methods that express a Linear Operator.  e.g. [pylops-distributed](https://github.com/equinor/pylops-distributed) seems to be taking this approach. However, the Linear Operator methods should take dask Arrays as arguments and return dask Arrays.
4. Otherwise, the ms2dirty and dirty2ms ducc methods can be wrapped with dask.array.blockwise calls representing dot and hdot for the gridder/degridder.

landmanbester commented:
I am not sure what is going wrong here. Some tests fail when I run them through pytest but not when I run them manually. Will keep digging

mreineck commented:
Do you use the `auto_choice` branch in CI? The `ducc0` branch will fail if you feed it `nu=nv=0`.

mreineck commented:
Unofortunately Github refuses to show me the test logs ...

sjperkins commented:
@landmanbester The test cases are failing on the base install (without optional packages) and test. You need to guard your ducc0 imports and use the `@requires_optional` decorator. I'd imagine you'd also need to configure the github workflow to install ducc0 for the complete install and test.

https://github.com/ska-sa/codex-africanus/blob/master/.github/workflows/ci.yml#L65-L72

https://github.com/ska-sa/codex-africanus/blob/master/.github/workflows/ci.yml#L65-L72

sjperkins commented:
> Unofortunately Github refuses to show me the test logs ...

@mreineck Out of interest, can you see https://github.com/ska-sa/codex-africanus/pull/204/checks?check_run_id=971810643?

And the following?

https://github.com/ska-sa/codex-africanus/runs/971809780

I'm never sure what github shows to maintainers compared to those with public access...

landmanbester commented:
Yes, sorry. I wasn't expecting the automated PR tests to pass. I mainly pushed it so I can discuss it with @sjperkins during the meeting tomorrow. Is there a way to push changes without triggering the tests or would I have to create a separate branch?

@mreineck I am using the master branch for now. I don't think it's anything in the gridder that's causing the tests to fail. They seem to fail on certain parametrisations using pytest but when I run the test manually with the same parameters they pass so I'm probably using @pmp wrong or something.

mreineck commented:
Hi @sjperkins, I can open the pages you have linked, but as soon as I try to open the sub-sections (e.g. by clicking on "run base test suite") I get "We are currently unable to download the log. Please try again later.". So it looks more like a technical problem.

mreineck commented:
Ahh, my fault ... my uMatrix was too paranoid. Sorry for the noise!

sjperkins commented:
> Yes, sorry. I wasn't expecting the automated PR tests to pass. I mainly pushed it so I can discuss it with @sjperkins during the meeting tomorrow. Is there a way to push changes without triggering the tests or would I have to create a separate branch?

Ah don't worry if they're failing: Its only important to get the green ticks before we press the merge button and the runs are short enough.

You can put "[skip ci]" in your commits but that will only skip the "push" event. There's another "pull_request" event but for various reasons it just runs on every commit.

landmanbester commented:
Alright, we have narrowed this down a bit. The problem only shows up when trying to process multiple chunks in parallel and this does not always happen. For example, I can trigger a failure manually about one in every hundred runs 

https://github.com/landmanbester/codex-africanus/blob/e3fcd1c68945888c31bc5696365cd666f5d6efb5/africanus/gridding/wgridder/tests/test_wgridder.py#L356

In this example, I am running multiple chunks in parallel but only giving the ducc gridder a single thread at a time. The failures become more frequent if I also give the ducc gridder multiple threads to work with. It does not seem to happen if I use the single-threaded scheduler, regardless of the number of threads I pass to the ducc gridder. This is a little bit out of my league but I think it points to possible issues with the intercommunication between threads when running multiple instances of the gridder. I will try to eliminate dask from the equation and come up with a minimal reproducible example. In the meantime, any suggestions are welcome. This is with the latest version of ducc0 on pypi.

mreineck commented:
I'm not sure what `dask` is doing exactly when it is mutithreading.

What ducc does is the following: it holds a global thread pool containing as many workers as there are virtual cores on the system. When executing a parallel region, the work is distributed among the requested number of these threads.

There may be some bad interaction between dask and ducc if dask is messing with the thread pool simultaneously from several threads. Not sure whether this can be fixed by mutexes ... recursive multithreading is a bit out of my league as well :(

sjperkins commented:
> I'm not sure what dask is doing exactly when it is mutithreading.

I'm going to allege that it's call the ducc gridder in multiple independent threads on independent data.

> What ducc does is the following: it holds a global thread pool containing as many workers as there are virtual cores on the system. When executing a parallel region, the work is distributed among the requested number of these threads.

> There may be some bad interaction between dask and ducc if dask is messing with the thread pool simultaneously from several threads. Not sure whether this can be fixed by mutexes ... recursive multithreading is a bit out of my league as well :(

I think re-entrancy is only an issue when functions need to call themselves?

I wonder if a lock is needed to ensure that only one thread can create and destroy the global threadpool? Maybe a singleton (although a bit of an anti-pattern these days) would work?

https://gitlab.mpcdf.mpg.de/mtr/ducc/-/blob/ducc0/src/ducc0/infra/threading.cc#L181-184

mreineck commented:
Thanks, that could be worth a try!

sjperkins commented:
I see C++11 introduced ``call_once`` which looks much better than a singleton!

mreineck commented:
... nope, that won't help, since other threads might still hold references to the pool.
Maybe it was a bad idea to make the pool resizable... the reason was that if you have a machine with, say, 24 hyperthreads and want to run a gridder on only 4 of them, then a full-sized thread pool will assign tasks more or less in a round robin fashion, which causes loads to jump around on the machine very quickly. In other words, whenever a new parallel area is reached, work will not be assigned to the 4 threads which just computed the last parallel region, but to 4 other ones.

Any suggestions?

sjperkins commented:
> ... nope, that won't help, since other threads might still hold references to the pool.
> Maybe it was a bad idea to make the pool resizable... the reason was that if you have a machine with, say, 24 hyperthreads and want to run a gridder on only 4 of them, then a full-sized thread pool will assign tasks more or less in a round robin fashion, which causes loads to jump around on the machine very quickly. In other words, whenever a new parallel area is reached, work will not be assigned to the 4 threads which just computed the last parallel region, but to 4 other ones.
> Any suggestions?


How about a [Multiton](https://en.wikipedia.org/wiki/Multiton_pattern)? This would involve creating one pool for each new number of threads requested. Assuming the user requests a constant number of threads per application, one should end up with a single threadpool in the general case. i.e. something like:

```c++
pool4 = get_pool(4);
pool8 = get_pool(8);
```

mreineck commented:
Sounds good, thanks! This still requires some careful locking, I guess, so I'll probably get it wrong the first few times :-)
Once this works, it will solve Landman's most urgent problem, i.e. there should be no more crashes.
However, if two Dask threads now both ask for a thread pool with, say,  4 threads, they will get the same object, and there will only be a 4-way parallelization overall. This is probably not what's intended...

sjperkins commented:
> Sounds good, thanks! This still requires some careful locking, I guess, so I'll probably get it wrong the first few times :-)

I'd suggest that a [Double-locking](https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_C++11) pattern might be the first approach to try.

> Once this works, it will solve Landman's most urgent problem, i.e. there should be no more crashes.
> However, if two Dask threads now both ask for a thread pool with, say, 4 threads, they will get the same object, and there will only be a 4-way parallelization overall. This is probably not what's intended...

Point taken. This sounds very familiar to the pattern @JSKenyon is using for numba and dask for calibration (numba has it's own internal threadpool) and it seems to work well. Jon can you elaborate on what you've seen in the numba context when you've used this pattern?

mreineck commented:
OK, I give up (at least for the moment, until I have something rock solid). I have now reverted to the non-resizable thread pool on the `auto-choice` branch. This should not break even when accessed from multiple threads simultanenously.
@landmanbester it would be great if you could throw your nastiest test cases at this!

landmanbester commented:
Thanks, @mreineck , I will do just that

landmanbester commented:
The plot thickens. I was still getting discrepancies on the auto-choice branch so I set up a test using concurrent.futures.ThreadPoolExecuter with manual chunking over row and assuming that all channels are gridded together into one MFS image. These always run through without issue and it seems that the dask version also runs through if I combine all channels into a single image. However, as soon as I try to split the data into multiple imaging bands and chunk over those as well I run into discrepancies. I suspect it is because of the way I am manually setting up the frequency mapping. Should probably rather use map_blocks for this

sjperkins commented:
> I suspect it is because of the way I am manually setting up the frequency mapping. Should probably rather use map_blocks for this

Are you sure you're not just selecting out the first frequency chunk? If there's a contraction of the frequency axis and there are multiple chunks, then they'll need to all be handled?

I think I've probably started a bad habit by doing this kind of thing -- it assumes a single chunk over the contraction axis.

```python
def _wrapper(..., freq, ...):
   freq = freq[0]
```

sjperkins commented:
> Should probably rather use map_blocks for this

I suspect blockwise is the correct tool to use here: map_blocks is too simple IMO.

landmanbester commented:
> Are you sure you're not just selecting out the first frequency chunk? If there's a contraction of the frequency axis and there are multiple chunks, then they'll need to all be handled?
> 
> I think I've probably started a bad habit by doing this kind of thing -- it assumes a single chunk over the contraction axis.
> 
> ```python
> def _wrapper(..., freq, ...):
>    freq = freq[0]
> ```

No, I am not. This also wouldn't explain why it only fails some of the time. The frequency mapping from nchan -> nband is implemented by keeping track of the starting indices in each band, as well as the number of channels falling in each band. Thus, I need to reset the indices to zero when chunking over frequency with something like
```
    freq_bin_idx -= freq_bin_idx.min()  # adjust for chunking
```
I was wondering if this logic might be faulty when chunking over row and frequency but now I do not see how that can be the case. I am pretty sure that it is the frequency chunking though. When I run the tests I always end up with a fraction of nband/nchan not agreeing between the two dirty images. Inspecting the images I see that it is always one imaging band that is incorrect. Sometimes this band is completely empty, other times the numerical values simply don't agree. I am not sure what could be causing this. It looks like the failures become more frequent with increasing thread over-subscription but it also happens when the number of chunks times the number of threads given to the gridder exactly match the number of virtual cores on my machine. This is a little bit of a mystery and I am not sure exactly how to debug it. I will try to implement both row and channel chunking manually using the ThreadPoolExecuter to see if I can recreate the issue that way

mreineck commented:
I  don't know why, but this somehow feels as if an array is expected to zero-intialized, but is not guaranteed to be. If you get the memory from the array from the OS, the contents will always be zero, but if it is constructed on recycled heap space, it will contain garbage. Depending on machine weather this can change from run to run.

I can't rule out that this happens inside `wgridder`, but I would not expect it at the moment, as I have never seen this behaviour in my tests so far.

sjperkins commented:
> I don't know why, but this somehow feels as if an array is expected to zero-intialized, but is not guaranteed to be. If you get the memory from the array from the OS, the contents will always be zero, but if it is constructed on recycled heap space, it will contain garbage. Depending on machine weather this can change from run to run.
> 
> I can't rule out that this happens inside `wgridder`, but I would not expect it at the moment, as I have never seen this behaviour in my tests so far.

In Python, I've seen this happy with `np.empty` calls.

landmanbester commented:
I do explicitly initialise the array to zeros here

https://github.com/landmanbester/codex-africanus/blob/5b34cceb25203888cf15602462fe81b802f6211f/africanus/gridding/wgridder/vis2im.py#L14

but I don't pass it into the gridder. Rather, I just assign the output of the gridder to it here

https://github.com/landmanbester/codex-africanus/blob/5b34cceb25203888cf15602462fe81b802f6211f/africanus/gridding/wgridder/vis2im.py#L17

We could possibly test this if there was an option to supply the output array to the gridder. Wasn't this possible at some stage @mreineck? Still, I do not see how the imaging bands could come out completely empty if it was just a case of the initial array containing garbage so I don't think this is the case

landmanbester commented:
Ha! It was the in-place modification of the freq_bin_idx array. Changing 

```
freq_bin_idx -= freq_bin_idx.min()  # adjust for chunking
```

to 

```
freq_bin_idx2 = freq_bin_idx - freq_bin_idx.min()  # adjust for chunking
```

fixes the issue. I am not sure how or why this fixes the issue but at least I can move on now

mreineck commented:
I removed the option of passing the output arrays in to make the interface simpler. That can be changed if necessary, but at least from a performance viewpoint I don't see much benefit.

Does the problem occur with or without enabled w-gridding? If it is enabled, I definitely fill the array with zeros; otherwise, I'm not yet completely sure, but I can add an extra zeroing step just to see if it changes anything on your end.

mreineck commented:
> Ha! It was the in-place modification of the freq_bin_idx array.

What??? You are saying that the right hand side of
```
freq_bin_idx -= freq_bin_idx.min()  # adjust for chunking
```
is not completely evaluated before the the modification of the left hand side starts?
What data type is `freq_bin_idx`?
If that is really true, I wouldn't touch whatever library is responsible for this with a 10 foot pole :-(

sjperkins commented:
> > Ha! It was the in-place modification of the freq_bin_idx array.
> 
> What??? You are saying that the right hand side of
> 
> ```
> freq_bin_idx -= freq_bin_idx.min()  # adjust for chunking
> ```
> 
> is not completely evaluated before the the modification of the left hand side starts?
> What data type is `freq_bin_idx`?
> If that is really true, I wouldn't touch whatever library is responsible for this with a 10 foot pole :-(

Interesting... Do you have a parallel decorator around the jitted function @landmanbester?

If so it probably needs to be something like 

```python
freq_bin_idx[:] -= freq_bin_idx.min()
```

as described here: https://numba.readthedocs.io/en/stable/user/parallel.html#explicit-parallel-loops

landmanbester commented:
> I removed the option of passing the output arrays in to make the interface simpler. That can be changed if necessary, but at least from a performance viewpoint I don't see much benefit.
> 
> Does the problem occur with or without enabled w-gridding? If it is enabled, I definitely fill the array with zeros; otherwise, I'm not yet completely sure, but I can add an extra zeroing step just to see if it changes anything on your end.

I have been testing without w-stacking enabled but it is clear now that this is not the issue so no need to alter the interface

mreineck commented:
Hmm, perhaps the problem is that some other code also holds a reference to `freq_bin_idx` and relies on it not changing its value ... that would be a saner explanation.

mreineck commented:
... or several threads do this modification concurrently on one and the same object => Instant mess of epic proportions :P

sjperkins commented:
Oh wait, the issues in the dask code where the input gets mutated. Ignore my comments on numba!

landmanbester commented:
> Interesting... Do you have a parallel decorator around the jitted function @landmanbester?
> 
> If so it probably needs to be something like
> 
> ```python
> freq_bin_idx[:] -= freq_bin_idx.min()
> ```
> 
> as described here: https://numba.readthedocs.io/en/stable/user/parallel.html#explicit-parallel-loops

The functions are not jitted so it can't be that. I think I see where it is going wrong. Each row chunk receives the same freq_bin_idx array and since multiple row chunks are processed concurrently the arrays get modified multiple times. I can explain it better at the meeting tomorrow @sjperkins

landmanbester commented:
> ... or several threads do this modification concurrently on one and the same object => Instant mess of epic proportions :P

Indeed!

landmanbester commented:
I've run into another issue after fixing the freq_mapping bug. I was still seeing discrepancies between dirty images computed with the serial vs. parallel chunking implementations so I decided to remove dask from the equation and reproduce the issue with concurrent.futures.ThreadPoolExecuter here

https://github.com/landmanbester/codex-africanus/blob/ef7e7fea285f1f9a3f4bc1768feaaba42fdd515c/africanus/gridding/wgridder/tests/test_wgridder.py#L384

In this test, I am computing the dirty image by chunking over row and processing multiple tasks in parallel. I then compare the result to that obtained when computing everything in one go. Each chunk calls ms2dirty resulting in multiple gridder instances, each of which potentially has multiple threads. If I run each gridder instance with a single thread then the results always agree. However, if I give each gridder instance multiple threads, I only get the same result if I split the number of rows into evenly sized chunks. In the example I point to above, 10000 rows are split into chunks of (3333, 3333, 3334) and the test fails if nthreads > 1. However, if I split the rows into (2500, 2500, 2500, 2500) chunks, the test passes regardless of the number of threads I pass to the gridder. Any idea what could be going wrong here @mreineck @sjperkins?

mreineck commented:
I can reproduce the problem. No idea what it is yet...

mreineck commented:
Problem most likely on my side. Stay tuned!

mreineck commented:
OK, should be fixed. Sorry, this was a really stupid bug; and I'm still extremely surprised that our tests didn't trigger it.

landmanbester commented:
Yes, that works now. Thanks a lot for the quick fix @mreineck

landmanbester commented:
Ah, ok most tests are passing. I was running into dtype issues when using single precision. I am not sure if this is a bug or a feature @mreineck but if you explicitly convert freq and uvw to single precision the gridder fails with error during array conversion

```
nx = 32, ny = 18, fov = 1.0, nrow = 1000, nchan = 5, nband = 1, epsilon = 1e-05, wstacking = True, nthreads = 1

    @pmp("nx", (32,))
    @pmp("ny", (18,))
    @pmp("fov", (1.0,))
    @pmp("nrow", (1000,))
    @pmp("nchan", (5,))
    @pmp("nband", (1,))
    @pmp("epsilon", (1e-5,))
    @pmp("wstacking", (True,))
    @pmp("nthreads", (1,))
    def test_gridder(nx, ny, fov, nrow, nchan, nband,
                     epsilon, wstacking, nthreads):
        # run comparison against dft with a frequency mapping imposed
        from africanus.gridding.wgridder import vis2im
        if epsilon >= 5e-6:
            real_type = "f4"
            complex_type = "c8"
            to_float = np.float32
        else:
            real_type = "f8"
            complex_type = "c16"
            to_float = np.float64
        np.random.seed(420)
        cellx = fov*np.pi/180/nx
        celly = fov*np.pi/180/ny * 1.1  # to test different cell sizes
        f0 = 1e9
        freq = (f0 + np.arange(nchan)*(f0/nchan)).astype(real_type)
        uvw = ((np.random.rand(nrow, 3)-0.5) /
               (cellx*freq[-1]/lightspeed)).astype(real_type)
        vis = (np.random.rand(nrow, nchan)-0.5 + 1j *
               (np.random.rand(nrow, nchan)-0.5)).astype(complex_type)
        wgt = np.random.rand(nrow, nchan).astype(real_type)
        step = nchan//nband
        if step:
            freq_bin_idx = np.arange(0, nchan, step)
            freq_mapping = np.append(freq_bin_idx, nchan)
            freq_bin_counts = freq_mapping[1::] - freq_mapping[0:-1]
        else:
            freq_bin_idx = np.array([0], dtype=np.int16)
            freq_bin_counts = np.array([1], dtype=np.int16)
        dirty = vis2im(uvw, freq, vis, wgt, freq_bin_idx, freq_bin_counts,
                       nx, ny, to_float(cellx), to_float(celly), 2*nx, 2*ny,
>                      epsilon, nthreads, wstacking)

tests/test_wgridder.py:81:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
vis2im.py:33: in vis2im
    epsilon, nthreads, do_wstacking)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

uvw = array([[ -56.295486,  -14.342838,  -71.155716],
       [-119.420074,  112.42556 ,   39.61469 ],
       [ -45.035786, -...31404],
       [ -74.07694 ,   28.749523,   51.050797],
       [ -81.285095,  -63.580513,  -95.56822 ]], dtype=float32)
freq = array([1.0e+09, 1.2e+09, 1.4e+09, 1.6e+09, 1.8e+09], dtype=float32)
vis = array([[-0.02242081-0.44403833j, -0.27906638+0.16218199j,
        -0.30894932+0.14744346j, -0.22500029-0.34248152j,
  ...031789j ,
         0.1456997 +0.44483975j,  0.13517176-0.4693286j ,
        -0.09090125+0.2682565j ]], dtype=complex64)
weights = array([[0.9612569 , 0.00309998, 0.7996618 , 0.47835594, 0.14544043],
       [0.68671286, 0.90700954, 0.24592584, 0.619...23, 0.7166283 , 0.21419209],
       [0.31240284, 0.32314706, 0.05445105, 0.67766815, 0.4332022 ]],
      dtype=float32)
freq_bin_idx = array([0]), freq_bin_counts = array([5]), nx = 32, ny = 18, cellx = 0.0005454154, celly = 0.0010665901, nu = 64, nv = 36, epsilon = 1e-05, nthreads = 1, do_wstacking = True

    def _vis2im_internal(uvw, freq, vis, weights, freq_bin_idx, freq_bin_counts,
                         nx, ny, cellx, celly, nu, nv, epsilon, nthreads,
                         do_wstacking):
        # adjust for chunking
        # need a copy here if using multiple row chunks
        freq_bin_idx2 = freq_bin_idx - freq_bin_idx.min()
        nband = freq_bin_idx.size
        # the extra dimension is required to allow for chunking over row
        dirty = np.zeros((1, nband, nx, ny), dtype=weights.dtype)
        for i in range(nband):
            ind = slice(freq_bin_idx2[i], freq_bin_idx2[i] + freq_bin_counts[i])
            dirty[0, i] = ms2dirty(uvw=uvw, freq=freq[ind], ms=vis[:, ind],
                                   wgt=weights[:, ind], npix_x=nx, npix_y=ny,
                                   pixsize_x=cellx, pixsize_y=celly,
                                   nu=nu, nv=nv, epsilon=epsilon,
                                   nthreads=nthreads,
>                                  do_wstacking=do_wstacking)
E           RuntimeError:
E           ./src/ducc0/bindings/pybind_utils.h: 44 (pybind11::array_t<T> ducc0::detail_pybind::toPyarr(const pybind11::object&) [with T = double]):
E
E           Assertion failure
E           error during array conversion

vis2im.py:24: RuntimeError

``` 
Is this the intended behaviour? It works if I leave uvw and freq to their default dtypes

mreineck commented:
Yes, this is intended. The gridder expects `uvw` and `freq` be double precision unconditionally. The other arrays (`ms`, `dirty`, and `wgt`) can be single or double precision (but their types must be consistent), and depending on the input types you'll get either single or double precision output.

Having `uvw` and `freq` always in double precision was an attempt to mimic the measurement sets: so far I never saw a MS where these quantities were single precision. But it's possible to adjust this, if necessary.

landmanbester commented:
Ok, thanks. Good toknow

mreineck commented:
You don't have to adjust `cellx` and `celly` either. They will be automatically re-converted to `double` when passed through `pybind11` :)

mreineck commented:
Hmmm, now that I think about it: are there any MSs that have double-precision DATA columns? If not, I can switch off a significant fraction of the gridding code, which should make compilation much faster. We can re-enable it for synthetic accuracy tests of course, but if we don't need it for everyday work, I'd say we disable gridding/degridding for double-precision MS data.

bennahugo commented:
It is permissible under the MSv2 specification, but I have not come across
it personally, no. Typically the digitized and channelized voltages have
far less precision than the 32bits assigned in typical datasets, so I don't
expect anyone to store visibility data as 64bit any time soon. I would keep
the grids themselves as double precision because accumulation onto the
grids is where most of the numerical error will arise. I've seen this
having an impact in the non-bda context for large data volumes.



On Thu, Aug 20, 2020 at 4:00 PM mreineck <notifications@github.com> wrote:

> Hmmm, now that I think about it: are there any MSs that have
> double-precision DATA columns? If not, I can switch off a significant
> fraction of the gridding code, which should make compilation much faster.
> We can re-enable it for synthetic accuracy tests of course, but if we don't
> need it for everyday work, I'd say we disable gridding/degridding for
> double-precision MS data.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ska-sa/codex-africanus/pull/204#issuecomment-677683759>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB4RE6XM6EDL6BRI5EGWBV3SBUT6NANCNFSM4OGY76MQ>
> .
>


-- 
--
Benjamin Hugo

PhD. student,
Centre for Radio Astronomy Techniques and Technologies
Department of Physics and Electronics
Rhodes University

Junior software developer
Radio Astronomy Research Group
South African Radio Astronomy Observatory
Black River Business Park
Observatory
Cape Town

landmanbester commented:
I usually convert everything to double-precision and just store the data in the same format as I read it in. I intend to test how big a difference this makes during imaging in pfb-clean but don't really have an answer for you now

sjperkins commented:
I think WSClean doesn't handle double precision complex data.

mreineck commented:
> Typically the digitized and channelized voltages have
far less precision than the 32bits assigned in typical datasets,

That's what I expected. Good to know this is valid at least for the time being.

> I would keep
the grids themselves as double precision because accumulation onto the
grids is where most of the numerical error will arise.

That's a point to keep in mind ... currently `wgridder` uses single precision grids when the input data are single precision. So far we haven't seen any issues, but if they ever turn up, we should try and see if they are caused by this choice of data type.

On the other hand I don't really see how numerical errors can accumulate; this would require having many visibilities very close together whose sum is much smaller than their individual values (without this kind of cancellation one can't get increasing relative errors, correct?). But would such a situation be physical?

Also I could imagine that codes like IDG would not work in this case, because I'm pretty sure they are not using double precision on the GPU :)

In any case, if this really is an issue, the necessary changes to `wgridder` are trivial; however, performance will drop by a factor of 2.

landmanbester commented:
That is strange. These tests all pass locally but not with the automated testing. Inspecting the logs I see there are typically a small number of elements in the array that don't match. I wonder what could be causing this. Any ideas @sjperkins ?

sjperkins commented:
> 
> 
> That is strange. These tests all pass locally but not with the automated testing. Inspecting the logs I see there are typically a small number of elements in the array that don't match. I wonder what could be causing this. Any ideas @sjperkins ?

You're installing ducc0 pypi release, not auto-choice branch?

landmanbester commented:
Ah yes, that could be it. So what should I do? Just skip the tests for now?

sjperkins commented:
> 
> 
> Ah yes, that could be it. So what should I do? Just skip the tests for now?

You could try overriding the install in `.github/workflows/ci.yml` similarly to way it's done with the nifty gridder.

mreineck commented:
`auto_choice` was merged and deleted today.

I'll try to make a new PyPI release soon!

landmanbester commented:
Success! Thanks for all your help with this @mreineck @sjperkins

sjperkins commented:
> Success! Thanks for all your help with this @mreineck @sjperkins

Green lights! A lot of work, but this is a very solid foundation. I wonder who should review? I'm willing, but it also might be good for @bennahugo to review too given his experience with gridding.

landmanbester commented:
I don't mind, you shouldn't need to really understand gridding since all of that happens in ducc. However, I think I did something wrong regarding making this an optional dependency. I also think I want to add the logic to allow for odd-sized images here instead of in the operator in pfb-clean. And I need to double-check that all the docs are correct. Maybe hold off for now so we can discuss at the meeting on Tuesday?

sjperkins commented:
> I don't mind, you shouldn't need to really understand gridding since all of that happens in ducc. However, I think I did something wrong regarding making this an optional dependency. I also think I want to add the logic to allow for odd-sized images here instead of in the operator in pfb-clean. And I need to double-check that all the docs are correct. Maybe hold off for now so we can discuss at the meeting on Tuesday?

Fine by me

mreineck commented:
It's no longer directly relevant to this PR, but there is now a wgridder branch called `new_thread_pool` which should show the slightly saner scheduling behaviour which I tried to implement before, but failed. @landmanbester it would be great if you could run your dask tests with this branch at some point!.

landmanbester commented:
I can confirm that all the wgridder tests pass with this branch locally

mreineck commented:
Thanks a lot for testing!

sjperkins commented:
@landmanbester and @mreineck I don't have any comments on the functionality as I think you've both put in a lot of work into testing it out.

One thing that I thought might be nice would be to keep the function naming consistent. The nifty wrappers use https://codex-africanus.readthedocs.io/en/latest/gridding-api.html#id1 the following conventions:

- grid
- dirty
- degrid
- model

where *dirty* and *model* perform the IFFT and FFT (can't remember which way round!).

@landmanbester mentioned that this may not be possible with the ducc0 interface because the FFT's are performed internally during the ms2dirty operation. Does the following hold?

```
F(G(v1) + G(v2) + G(v3) + ... + G(vn)) == F(G(v1)) + F(G(v2)) + F(G(v3)) + ... + F(G(vn))
```

where F is the FFT, G is the gridding operation and v1 ... vn are the separate visibility chunks.

If so, I never realised and perhaps the nifty wrapper conventions should change!

mreineck commented:
> @landmanbester mentioned that this may not be possible with the ducc0 interface because the FFT's are performed internally during the ms2dirty operation.

Correct. Doing all in one go opens up a lot more optimization potential, makes the interface simpler and can also save memory.
If you need the intermediate data products, that could prove quite tricky...

Also, the *dirty* and *model* operations do much more than simply an (I)FFT for wide-field imaging.

sjperkins commented:
> If you need the intermediate data products, that could prove quite tricky...

> Also, the dirty and model operations do much more than simply an (I)FFT for wide-field imaging.

I think my only concern is whether the grids from gridding multiple independent visibility chunks can be summed to produce the correct final image. This is how it's currently set up in dask.

mreineck commented:
Yes, that will definitely work. However, don't expect to reproduce "exactly" the same dirty image (i.e. up to numerical precision) with different subdivision strategies for the visibilities, as the code optimizes oversampling factors etc. to be s fast as possible, but stay within the requested error budget. The decisions it makes can depend on the uvw coverage of the visibilities.

mreineck commented:
In general you should see the best performance if you subdivide the set of visibilities such that each block has the smallest possible extent in w.

sjperkins commented:
> 
> 
> Yes, that will definitely work. However, don't expect to reproduce "exactly" the same dirty image (i.e. up to numerical precision) with different subdivision strategies for the visibilities, as the code optimizes oversampling factors etc. to be s fast as possible, but stay within the requested error budget. The decisions it makes can depend on the uvw coverage of the visibilities.

Great, this is perfect!

> In general you should see the best performance if you subdivide the set of visibilities such that each block has the smallest possible extent in w.

This might not be immediately possibly, given that our Measurement Sets are monotically ordered in TIME, but I think @landmanbester is considering a baseline re-ordering via https://github.com/pangeo-data/rechunker.

> [Edit] Also, the larger the number of visibilities in a single call, the better (as long as everything fits into memory).

👍

landmanbester commented:
> I think my only concern is whether the grids from calling ms2dirty on multiple independent visibility chunks can be summed to produce the correct dirty image. This is how it's currently set up in dask.

This is already being tested by comparison to the DFT (called explicit gridder in the test suite)

landmanbester commented:
> This might not be immediately possibly, given that our Measurement Sets are monotically ordered in TIME, but I think @landmanbester is considering a baseline re-ordering via https://github.com/pangeo-data/rechunker.

My current plan for the interface to the data is to read from a measurement set, compute a chunking strategy which I write to an intermediate format (thinking .zarr files here) and then write back to the MS right at the end. Since there is no reason to maintain the time ordering of the data during gridding we can in principle do whatever we want within this intermediate format. I was thinking of chunking per imaging band as long as the full band fits into memory. If not I was going to subdivide the band by row grouped by w to minimise the extent in w in each chunk. Is there anything else (like sorting the data for example) that would make the per chunk gridding more efficient? Note this strategy might not be very IO efficient since it assumes that only a single chunk is read into memory at any time. I suspect we will have to play with the number of dask workers vs. gridder threads etc. to see what works best in the end. Suggestions welcome. 

Regarding the naming convention, I do not feel too strongly about this but I think the names are pretty descriptive in terms of what they do. im2residim is a bit of a mouthful though...

mreineck commented:
I'd say that, depending on the number of visibilities and the image size + number of planes in a single call, you might not even need to worry about rearranging anything. The only advantage you have from this is a reduced number of FFTs, and if you have enough visibilities per call and the dirty images are not gigantic (i.e. beyond 10000x10000 or so), this should not make a big difference. I guess that you'll have to do a few experiments to get a feeling for the scaling; setting `verbosity` to 1 will now also produce a detailed timing analysis.

BTW, I have a revised version of the gridder in the pipeline which has a significantly reduced memory overhead. With that version, the additional memory allocated during gridding/degridding (in bytes, assuming single precision) will be
- space for one grid (`8*nxdirty*nydirty*ofactor**2`)
- possibly space for one additional dirty image (`4*nxdirty*nydirty`)
- space for a mask (`1*nrows*nchannels`)
- space for the indexing information (`16*alpha*nrows*nchannels`)

ofactor is dynamically chosen and will lie in the range [1.2;2]
alpha depends a lot on the way how the data is flagged: in the absolute worst case it can become 0.5 (that would be a completely pathological data set with every other channel flagged), but is typically way below 0.1.

On average this is much lower than what the current version uses; I expect to merge that within the next weeks.

I'm mentioning this because on a large memory node this version should allow you to process absolutely gigantic amounts of data in one go; on a 128GB machine, processing 10 billion visibilities per call should not be a problem. And in this regime, FFTs should not be dominant any more :)

landmanbester commented:
Great, thanks for clarifying. That will help me to plan my chunking strategy

sjperkins commented:
> I'm mentioning this because on a large memory node this version should allow you to process absolutely gigantic amounts of data in one go; on a 128GB machine, processing 10 billion visibilities per call should not be a problem. And in this regime, FFTs should not be dominant any more :)

It's going to be fascinating to try this out on large problem sizes. I suspect we might become I/O limited by the Measurement Set.

sjperkins commented:
@landmanbester Regarding the naming conventions, I personally favour the following:

- dirty
- model
- residual

I don't have anything against the X2Y convention. It is descriptive and clear.

My niggle is that codex has been built up on single noun/verb style functions and it would be good to keep the convention consistent (within codex) for now. Whether it's a good convention might be worth another discussion ;-)

mreineck commented:
> It's going to be fascinating to try this out on large problem sizes. I suspect we might become I/O limited by the Measurement Set.

Possibly :) On the other hand, this approach might reduce total I/O volume, because fewer passes over the Measurement Set are needed. If I compare `wsclean` runs with and without wgridder, the standard gridder always has to go several times through the data for larger MSs, while everything fits into RAM when using wgridder (at least for the MSs I have access to :P)

BTW, my "10 billion visibilities" number may have been a bit too optimistic, because I forgot that you most likely will use weights as well, which adds another 4 bytes per visibility ...

mreineck commented:
> I will make it the default behaviour for the gridder to choose its own nu, nv values i.e. set them to zero by default.

Good choice! From the next release on, these parameters will be ignored anyway (and whenever `ducc1` is released, they will be removed from the interface).

> Something I am less sure about is whether I should set a default value for the precision (epsilon). For single-precision, we can always set this to the max accuracy of 1e-5.

You can in principle go to 1e-7 now for single precision, but for most cases that will be way too much, and computation time tends to increase a lot if you go to really low epsilons.
FWIW, `wsclean` uses a hard-wired epsilon of 1e-4 for wgridder, and Andre claims that this should be more than sufficient for any application he is aware of.
He measured the resulting accuracy in a similar way to https://arxiv.org/abs/1908.11232 and got the plot below. Maybe that is helpful to decide on what you should use.

> However, what do we do about double precision where the maximum precision of 2e-13 is usually complete overkill. Set it to something sensible like 1e-8?

What will be the application scenario for this? If you don't have input data with this kind of relative accuracy, how can such a small epsilon help you?

landmanbester commented:
> What will be the application scenario for this? If you don't have input data with this kind of relative accuracy, how can such a small epsilon help you?

I have no idea. I've not tested how big a difference this makes during imaging. I usually convert everything to double precision and set epsilon to 1e-7. This is just an attempt to avoid accumulation errors but if wsclean can get away with 1e-4 I can probably save a lot of memory. Still would like to have the option in the gridder in case we ever get double precision measurement sets. I'll set the default to 1e-7 when using double precision

mreineck commented:
OK. As I said, I don't know enough to judge how serious the accumulation errors could become. It's probably worth to test this for every big job. Single precision not only uses less memory, but is typically also more than twice as fast.

landmanbester commented:
> OK. As I said, I don't know enough to judge how serious the accumulation errors could become. It's probably worth to test this for every big job. Single precision not only uses less memory, but is typically also more than twice as fast.

Thanks, I'll try this out

landmanbester commented:
Hmmm not sure what's going wrong. I just ran the tests locally with ducc0 from PYPI and they all pass. The failures are all due to a lack of expected relative precision (1.5e-5 as apposed to 1e-5 at single precision). Not a big deal, I'll just lower the precision but would be good to know whats going on

sjperkins commented on africanus/gridding/wgridder/dask.py:
```diff
@@ -0,0 +1,211 @@
+
+    # set precision
+    if epsilon is None:
+        if image[0, 0].dtype == np.float64:
```
Would it be possible to avoid indexing and use `image.dtype` here ?

landmanbester replied:
done

sjperkins commented on africanus/gridding/wgridder/dask.py:
```diff
@@ -0,0 +1,211 @@
+
+    # set precision
+    if epsilon is None:
+        if vis[0, 0].dtype == np.complex128:
```
Would it be possible to avoid indexing and use image.dtype here ?

sjperkins commented on africanus/gridding/wgridder/im2vis.py:
```diff
@@ -0,0 +1,147 @@
+          flag=None, celly=None, epsilon=None, nthreads=1, do_wstacking=True):
+    # set precision
+    if epsilon is None:
+        if type(image[0, 0, 0]) == np.float64:
```
Would it be possible to avoid indexing and use image.dtype here ?

sjperkins commented on africanus/gridding/wgridder/im2vis.py:
```diff
@@ -0,0 +1,147 @@
+    denotes visibilities of shape :code:`(row, chan)` and
+    :math:`x` is the image of shape :code:`(band, nx, ny)`.
+
+    The number of imaging bands :code:`(band)` is has to
```
```suggestion
    The number of imaging bands :code:`(band)` has to
```

sjperkins commented on africanus/gridding/wgridder/im2vis.py:
```diff
@@ -0,0 +1,147 @@
+
+    There is an option to provide weights during degridding.
+    This option is made available to allow for the possibility
+    of having self adjoint gridding and degridding operators.
```
How about?

```
There is an option to provide weights during degridding
to cater for self adjoint gridding and degridding operators.
```

landmanbester replied:
Done

sjperkins commented on africanus/gridding/wgridder/tests/test_wgridder.py:
```diff
@@ -0,0 +1,403 @@
+                                                     np.sum(np.abs(b)**2)))
+
+
+def explicit_gridder(uvw, freq, ms, wgt, nxdirty, nydirty, xpixsize, ypixsize,
```
It would be nice to jit this function to speed up tests (if the dimensions are large).

landmanbester replied:
Get the following when I try this
```
E           NotImplementedError: Failed in nopython mode pipeline (step: analyzing bytecode)
E           CALL_FUNCTION_EX with **kwargs not supported
```
I guess this is because of the indexing='ij' inside np.meshgrid. This function is only called a handful of times (only for test_gridder()) and the tests don't take long. I don't think it will really make a big difference

sjperkins commented on africanus/gridding/wgridder/tests/test_wgridder.py:
```diff
@@ -0,0 +1,403 @@
+    # should agree to within epsilon
+    rmax = np.maximum(np.abs(residim_np).max(), np.abs(residim_da).max())
+    assert_array_almost_equal(
+        residim_np/rmax, residim_da/rmax, decimal=decimal)
```
These tests look very comprehensive.

landmanbester replied:
I went a bit overboard on the parametrisations. Still, the full test suite for the wgridder runs in about 50s on my laptop

sjperkins commented on africanus/gridding/wgridder/vis2im.py:
```diff
@@ -0,0 +1,149 @@
+    # need a copy here if using multiple row chunks
+    freq_bin_idx2 = freq_bin_idx - freq_bin_idx.min()
+    nband = freq_bin_idx.size
+    if type(vis[0, 0]) == np.complex64:
```
Maybe use vis.dtype?

sjperkins commented on africanus/gridding/wgridder/vis2im.py:
```diff
@@ -0,0 +1,149 @@
+          do_wstacking=True):
+    # set precision
+    if epsilon is None:
+        if type(vis[0, 0]) == np.complex128:
```
vis.dtype

sjperkins commented on africanus/gridding/wgridder/vis2im.py:
```diff
@@ -0,0 +1,149 @@
+    :math:`V` denotes visibilities of shape :code:`(row, chan)` and
+    :math:`I^D` is the dirty image of shape :code:`(band, nx, ny)`.
+
+    The number of imaging bands :code:`(band)` is has to
```
```suggestion
    The number of imaging bands :code:`(band)` must
```

sjperkins commented on africanus/gridding/wgridder/vis2im.py:
```diff
@@ -0,0 +1,149 @@
+    :code:`(chan)` at which the data were obtained.
+    The mapping from :code:`(chan)` to :code:`(band)` is described
+    by :code:`freq_bin_idx` and :code:`freq_bin_counts` as
+    described below.
```
```
The mapping from :code:`(chan)` to :code:`(band)` is described
by :code:`freq_bin_idx` and :code:`freq_bin_counts` below.
```

sjperkins commented on africanus/gridding/wgridder/vis2im.py:
```diff
@@ -0,0 +1,149 @@
+    by :code:`freq_bin_idx` and :code:`freq_bin_counts` as
+    described below.
+
+    Note that, if self adjoint gridding and degridding opeartors
```
```suggestion
    Note that, if self adjoint gridding and degridding operators
```

sjperkins commented on africanus/gridding/wgridder/vis2im.py:
```diff
@@ -0,0 +1,149 @@
+    described below.
+
+    Note that, if self adjoint gridding and degridding opeartors
+    are required then :code:`weights` should actually be the square
```
```suggestion
    are required then :code:`weights` should be the square
```

sjperkins commented on africanus/gridding/wgridder/vis2im.py:
```diff
@@ -0,0 +1,149 @@
+    Note that, if self adjoint gridding and degridding opeartors
+    are required then :code:`weights` should actually be the square
+    root of what is typically referred to as imaging weights and
+    the same weights need to be passed into the degridder.
```
```suggestion
    should also be passed into the degridder.
```

sjperkins commented on africanus/gridding/wgridder/vis2im.py:
```diff
@@ -0,0 +1,149 @@
+    are required then :code:`weights` should actually be the square
+    root of what is typically referred to as imaging weights and
+    the same weights need to be passed into the degridder.
+    In this case, the data that are passed in need be pre-whitened.
```
```suggestion
    In this case, the data needs to be pre-whitened.
```

sjperkins commented on africanus/gridding/nifty/dask.py:
```diff
@@ -105,7 +105,7 @@ def _nifty_indices(baselines, grid_config, flag,
 def _nifty_grid(baselines, grid_config, indices, vis, weights):
     """ Wrapper function for creating a grid of visibilities per row chunk """
-    assert len(vis) == 1 and type(vis) == list
+    assert len(vis) == 1 and isinstance(vis, list)
```
```suggestion
    assert len(vis) == 1 and type(vis) is list
```

sjperkins commented on africanus/gridding/wgridder/im2residim.py:
```diff
@@ -0,0 +1,167 @@
+    denotes visibilities of shape :code:`(row, chan)` and
+    :math:`x` is the image of shape :code:`(band, nx, ny)`.
+
+    The number of imaging bands :code:`(band)` is has to
```
```suggestion
    The number of imaging bands :code:`(band)` must
```

sjperkins commented on africanus/util/dask_util.py:
```diff
@@ -39,10 +39,10 @@ def format_time(t):
 
 def key_bin(key):
-    if type(key) == tuple:
+    if isinstance(key, tuple):
```
```suggestion
    if type(key) is tuple:
```

sjperkins commented on africanus/util/dask_util.py:
```diff
@@ -39,10 +39,10 @@ def format_time(t):
         key = key[0]
 
-    if type(key) == bytes:
+    if isinstance(key, bytes):
```
```suggestion
    if type(key) is bytes:
```

sjperkins commented on africanus/util/testing.py:
```diff
@@ -25,7 +25,7 @@ def in_pytest():
 def mark_in_pytest(in_pytest=True):
     """ Mark if we're in a pytest run """
-    if not type(in_pytest) == bool:
+    if not isinstance(in_pytest, bool):
```
```suggestion
    if type(in_pytest) is not bool:
```

sjperkins reviewed (changes requested):
Looks very good. I've some minor comments and suggested some typo corrections.

sjperkins commented:
Perhaps this can be a birthday PR ;-)

mreineck commented:
> That is a bit strange. These tests all pass locally on the ducc0 branch. Is the PYPI release up to date @mreineck?

Not completely, but I'm not aware of a bug fix since the last release.
I'll try to make another release soon, but before that I have to apply another patch that fixes a rare source of deadlocks with the new thread pool implementation. Might take a few days.

landmanbester commented:
I don't recall making the changes to africanus/util gridding/nifty but I've implemented the suggestions. I see numba is still pinned to <= 0.49 in setup.py. Maybe a result of merging in master when the perleypolyhedron gridder was in there. I take it we don't want to keep it as is?

mreineck commented:
FYI: I just released `ducc0` 0.6, which contains the promised memory reductions; it's a bit faster as well!

sjperkins commented:
> FYI: I just released `ducc0` 0.6, which contains the promised memory reductions; it's a bit faster as well!

Thanks @mreineck. I've opened https://github.com/ska-sa/codex-africanus/issues/221 to track this.
